src/dd/Approximation.cpp

```diff
@@ -0,0 +1,246 @@
+#include "dd/Node.hpp"
+#include "dd/Package.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <forward_list>
+#include <limits>
+#include <queue>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+namespace dd {
+namespace {
+constexpr uint16_t FLAG_DELETE = 0b100;
+constexpr uint16_t FLAG_LEFT = 0b010;
+constexpr uint16_t FLAG_RIGHT = 0b001;
+
+/**
+ * @brief A node-contribution pair and distance value for prioritization.
+ */
+class LayerNode {
+public:
+  explicit LayerNode(vNode* node)
+      : ptr(node), contribution(1.),
+        distance(std::numeric_limits<double>::max()) {}
+
+  LayerNode(vNode* node, double contribution, double budget)
+      : ptr(node), contribution(contribution),
+        distance(std::max<double>(0, contribution - budget)) {}
+
+  bool operator<(const LayerNode& other) const {
+    if (distance == 0 && other.distance == 0) {
+      return contribution < other.contribution;
+    }
+    return distance > other.distance;
+  }
+
+  vNode* ptr;
+  double contribution;
+
+private:
+  double distance;
+};
+
+/// @brief Priority queue of nodes, sorted by their distance to a given budget.
+using Layer = std::priority_queue<LayerNode>;
+/// @brief Maps nodes to their respective contribution.
+using Contributions = std::unordered_map<vNode*, double>;
+/// @brief Maps old nodes to rebuilt edges.
+using Lookup = std::unordered_map<const vNode*, vEdge>;
+/// @brief A node, a flag indicating left or right, and its contribution.
+using Terminal = std::tuple<vNode*, uint16_t, double>;
+/// @brief List of potentially deletable terminal edges.
+using TerminalList = std::forward_list<Terminal>;
+
+/**
+ * @brief Remove terminals with a contribution above the new budget.
+ */
+TerminalList removeAbove(const TerminalList& old, const double budget) {
+  TerminalList upd{};
+  for (const auto& candidate : old) {
+    if (budget >= std::get<2>(candidate)) {
+      upd.emplace_front(candidate);
+    }
+  }
+  return upd;
+}
+
+/**
+ * @brief Search and mark nodes for deletion until the budget 1 - @p fidelity is
+ * exhausted.
+ * @details Uses a prioritized iterative-deepening search.
+ * Iterating layer by layer ensures that each node is only visited once.
+ */
+std::pair<double, Qubit> mark(VectorDD& state, const double fidelity,
+                              struct ApproxMeta* meta) {
+  Layer curr{};
+  curr.emplace(state.p);
+
+  double budget = 1 - fidelity;
+  Qubit min{std::numeric_limits<Qubit>::max()};
+
+  TerminalList candidates{};
+
+  ApproxMeta rmeta{};
+
+  while (budget > 0) {
+    Contributions c; // Stores contributions of the next layer.
+    while (!curr.empty()) {
+      const LayerNode n = curr.top();
+      curr.pop();
+      rmeta.nodesVisited++;
+
+      // If possible, flag a node for deletion and decrease the budget.
+      // If necessary, reset the lowest qubit number effected.
+      if (budget >= n.contribution) {
```

Comment:
Maybe that's just my brain being wired wrongly, but somehow the order in these comparisons triggers me.
I'd find it more "logical"/natural for all of these to read
If (contribution <= budget)

Maybe you agree.